Reset _prev_elem cache when mesh changes

lindsayad opened the pull request:
There is a non-zero chance that we will call `initDofIndices()` with the
same element before and after adaptivity when its dof indices have
changed, so we must make sure to reset the `_prev_elem` cache in order
to make sure the dof indices get re-initialized

Refs libmesh/libmesh#2963

roystgnr commented:
Oh, yikes!  Let me guess - in the adaptivity case, `_prev_elem` was getting coarsened away and a new different Elem was getting allocated in its place?

I remember running into a similar caching bug ... in FE or FEMContext? ... years ago.  Horribly hard to catch.  This was the sort of thing that motivated me to want to run tests on 1 through N processors (for N as large as possible) in the first place.

lindsayad commented:
> Oh, yikes! Let me guess - in the adaptivity case, `_prev_elem` was getting coarsened away and a new different Elem was getting allocated in its place?

I didn't verify, but I hypothesized that it was something like that. Too bad we don't have matching np count for MOOSE test for CI :-/

lindsayad commented:
@rwcarlsen can I get your opinion on this? If you don't like this we could perhaps do something with `unique_id()` instead?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/784118/) on e899ca5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18448/site)

This comment will be updated on new commits.

lindsayad commented:
Hmm well there you go... looks like the caching is pretty important! Bunch of timeouts now

roystgnr commented:
If you want to try comparing `unique_id()`, that should be safe even after adaptivity.

lindsayad commented:
Just wanted to get a profile before making the change to `unique_id`. Sure enough, the `initDofIndices` call-path shows up
![no-caching](https://user-images.githubusercontent.com/10248304/127057168-28942a55-f111-4bd3-be95-12fbe3ae9f06.png)

rwcarlsen commented:
This is not going to work well - as the profile shows.  I wrote the code like this in order to make is more robust in the first place.  Previously, users of the variable API had to try to remember whether or not the dof indices were init'd/updated correctly - and call initDofIndices manually if they weren't.  I rewrote it so that we lazily updated it whenever anyone tried to retrieve the dof indices.  I think the right approach is just to more robustly track whether or not they dof indices are "dirty" - so we know when they need to be updated. Just init'ing them over and over is yuck.

rwcarlsen reviewed (changes requested):
This is not going to work well - as the profile shows. I wrote the code like this in order to make is more robust in the first place. Previously, users of the variable API had to try to remember whether or not the dof indices were init'd/updated correctly - and call initDofIndices manually if they weren't. I rewrote it so that we lazily updated it whenever anyone tried to retrieve the dof indices. I think the right approach is just to more robustly track whether or not they dof indices are "dirty" - so we know when they need to be updated. Just init'ing them over and over is yuck.

lindsayad committed:
Reset _prev_elem cache when mesh changes

There is a non-zero chance that we will call `initDofIndices()` with the
same element before and after adaptivity when its dof indices have
changed, so we must make sure to reset the `_prev_elem` cache in order
to make sure the dof indices get re-initialized

Refs libmesh/libmesh#2963

lindsayad commented:
Ignore the fact that I'm a dummy about header inclusion... changing to `unique_id` I'm seeing the original behavior with 13 procs. I'm thinking that the same element winds up with different dof indices after adaptivity. I assume that's possible @roystgnr ?

lindsayad commented:
Yea the element is definitely the same but now it has different dof indices. This is gross

lindsayad commented:
Ok I think this is the right fix

rwcarlsen reviewed (approved):
Much better.

roystgnr commented:
> I assume that the same element having different dof indices after adaptivity is fully possible @roystgnr ?

Damn, of course it is.  And this is orders of magnitude more likely than my reallocation theory too.  Sorry to send you chasing after that red herring.
